sklearn/cluster/_optics.py

```diff
@@ -649,6 +649,14 @@ def _set_reach_dist(
     predecessor_[unproc[improved]] = point_index
 
 
+@validate_params(
+    {
+        "reachability": [None, np.ndarray],
+        "core_distances": [None, np.ndarray],
+        "ordering": [None, np.ndarray],
```

Comment:
It does not seem that `None` is an option for any of those parameters.

Reply:
Thanks for the feedback!

When I run ```pytest -vl sklearn/tests/test_public_functions.py```

with 

```
@validate_params(
    {
        "reachability": [np.ndarray],
        "core_distances": [np.ndarray],
        "ordering": [np.ndarray],
        "eps": [Interval(Real, 0, None, closed="neither")],
    }
)
```
I get the error:

```constraint = <sklearn.utils._param_validation._InstancesOf object at 0x10f0cb340>

    def generate_valid_param(constraint):
        """Return a value that does satisfy a constraint.
    
        This is only useful for testing purpose.
    
        Parameters
        ----------
        constraint : Constraint instance
            The constraint to generate a value for.
    
        Returns
        -------
        val : object
            A value that does satisfy the constraint.
        """
        if isinstance(constraint, _ArrayLikes):
            return np.array([1, 2, 3])
    
        if isinstance(constraint, _SparseMatrices):
            return csr_matrix([[0, 1], [1, 0]])
    
        if isinstance(constraint, _RandomStates):
            return np.random.RandomState(42)
    
        if isinstance(constraint, _Callables):
            return lambda x: x
    
        if isinstance(constraint, _NoneConstraint):
            return None
    
        if isinstance(constraint, _InstancesOf):
>           return constraint.type()
E           TypeError: ndarray() missing required argument 'shape' (pos 1)

constraint = <sklearn.utils._param_validation._InstancesOf object at 0x10f0cb340>

sklearn/utils/_param_validation.py:871: TypeError
```

If we just focus on one of the array parameters for an example, as far as I can tell the `reachability` array gets initialized with a call to ```np.empty()``` before being passed to `cluster_optics_dbscan`. Perhaps I am interpreting the error message incorrectly, but I am unclear as to a) why a `shape` argument would be required for the validation and b) how a `shape` argument would be provided in the `validate_params` dict if so (I didn't see an example of providing a `shape` argument in other PRs for this issue).

If you have any suggestions please let me know and I can do my best to fix this PR.

Reply:
this issue was fixed in another PR. Removing None should work now. I directly pushed the change